Sources/KSCrashRecording/Monitors/KSCrashMonitor_Memory.mm

```diff
@@ -0,0 +1,567 @@
+//
+//  KSCrashMonitor_NSException.m
+//
+//  Created by Karl Stenerud on 2012-01-28.
+//
+//  Copyright (c) 2012 Karl Stenerud. All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall remain in place
+// in this source code.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+#import "KSCrashMonitor_Memory.h"
+
+#import "KSCrash.h"
+#import "KSCrashC.h"
+#import "KSCrashMonitorContext.h"
+#import "KSCrashAppMemory.h"
+#import "KSID.h"
+#import "KSStackCursor.h"
+#import "KSStackCursor_SelfThread.h"
+#import "KSStackCursor_MachineContext.h"
+#import "KSCrashReportFields.h"
+
+#import <Foundation/Foundation.h>
+#import <mutex>
+#import <sys/time.h>
+#import <sys/mman.h>
+
+#import "KSLogger.h"
+
+#if TARGET_OS_IOS
+#import <UIKit/UIKit.h>
+#endif
+
+// ============================================================================
+#pragma mark - Globals -
+// ============================================================================
+
+static volatile bool g_isEnabled = 0;
+static volatile bool g_hasPostEnable = 0;
+
+// Install path for the crash system
+static NSURL *g_installURL = nil;
+
+// The memory tracker
+@class MemoryTracker;
+static MemoryTracker *g_memoryTracker = nil;
+
+// file mapped memory
+static std::mutex g_lock [[clang::no_destroy]];
```

Comment:
It might be worth avoiding C++ for synchronisation unless it's absolutely necessary here too.

Reply:
It is simpler since we can use scope and the stack to our advantage. We can do this:

```
{
std::lock_guard<std::mutex> l(g_lock);
//... use resource
}
```

Instead of this:

```
[g_lock lock];
//... use resource
[g_lock unlock];
```

There are ways we could use the cleanup attributes to do something similar to defer which could work, but it's still more complicated and harder to read than RAII.

Reply:
It seems like using these C locks is a good way to avoid C++. It's clear that the RAII approach simplifies things, but sticking with C locks can still be manageable.